File: app/src/main/java/com/nfcalarmclock/tts/NacTextToSpeechDialog.java
package com.nfcalarmclock.tts;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Bundle;
import android.view.View;
import android.widget.NumberPicker;
import android.widget.RelativeLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.checkbox.MaterialCheckBox;

import com.nfcalarmclock.R;
import com.nfcalarmclock.shared.NacSharedConstants;
import com.nfcalarmclock.shared.NacSharedPreferences;
import com.nfcalarmclock.util.dialog.NacDialogFragment;

import java.util.List;

/**
 */
public class NacTextToSpeechDialog
	extends NacDialogFragment
	implements View.OnClickListener
{

	/**
	 * Listener for when text-to-speech options are selected.
	 */
	@SuppressWarnings("UnnecessaryInterfaceModifier")
	public interface OnTextToSpeechOptionsSelectedListener
	{
		public void onTextToSpeechOptionsSelected(boolean useTts, int freq);
	}

	/**
	 * Tag for the class.
	 */
	public static final String TAG = "NacAlarmTextToSpeechDialog";

	/**
	 * Default flag indicating whether text-to-speech should be used or not.
	 */
	private boolean mDefaultUseTts;

	/**
	 * Default text-to-speech frequency.
	 */
	private int mDefaultTtsFrequency;

	/**
	 * Checkbox for setting whether text-to-speech should be used or not.
	 */
	private MaterialCheckBox mShouldUseTtsCheckBox;

	/**
	 * Summary text for whether text-to-speech should be used or not.
	 */
	private TextView mShouldUseTtsSummary;

	/**
	 * Scrollable picker to choose the text-to-speech frequency.
	 */
	private NumberPicker mTtsFrequencyPicker;

	/**
	 * Listener for when a text-to-speech option and frequency is selected.
	 */
	private OnTextToSpeechOptionsSelectedListener mOnTextToSpeechOptionsSelectedListener;

	/**
	 * Call the OnTextToSpeechOptionsSelectedListener object, if it has been set.
	 */
	public void callOnTextToSpeechOptionsSelectedListener()
	{
		OnTextToSpeechOptionsSelectedListener listener =
			this.getOnTextToSpeechOptionsSelectedListener();

		if (listener != null)
		{
			NumberPicker picker = this.getTtsFrequencyPicker();
			boolean useTts = this.shouldUseTts();
			int freq = picker.getValue();

			listener.onTextToSpeechOptionsSelected(useTts, freq);
		}
	}

	/**
	 * Get the default flag indicating whether text-to-speech should be used or
	 * not.
	 *
	 * @return The default flag indicating whether text-to-speech should be used
	 *     or not.
	 */
	public boolean getDefaultUseTts()
	{
		return this.mDefaultUseTts;
	}

	/**
	 * Get the default text-to-speech frequency.
	 *
	 * @return The default text-to-speech frequency.
	 */
	public int getDefaultTtsFrequency()
	{
		return this.mDefaultTtsFrequency;
	}

	/**
	 * Get the OnTextToSpeechOptionsSelectedListener object.
	 *
	 * @return The OnTextToSpeechOptionsSelectedListener object.
	 */
	public OnTextToSpeechOptionsSelectedListener getOnTextToSpeechOptionsSelectedListener()
	{
		return this.mOnTextToSpeechOptionsSelectedListener;
	}

	/**
	 * Get the checkbox for setting whether text-to-speech should be used or not.
	 *
	 * @return The checkbox for setting whether text-to-speech should be used or not.
	 */
	private MaterialCheckBox getShouldUseTtsCheckBox()
	{
		return this.mShouldUseTtsCheckBox;
	}

	/**
	 * Get the ummary text for whether text-to-speech should be used or not.
	 *
	 * @return The summary text for whether text-to-speech should be used or not.
	 */
	private TextView getShouldUseTtsSummary()
	{
		return this.mShouldUseTtsSummary;
	}

	/**
	 * Get the scrollable picker for the text-to-speech frequency.
	 *
	 * @return The scrollable picker for the text-to-speech frequency.
	 */
	private NumberPicker getTtsFrequencyPicker()
	{
		return this.mTtsFrequencyPicker;
	}

	/**
	 * Called when the check box is clicked.
	 */
	@Override
	public void onClick(View view)
	{
		int id = view.getId();

		if (id == R.id.should_use_tts)
		{
			this.toggleShouldUseTts();
			this.setupShouldUseTtsSummary();
			this.setupTtsFrequencyEnabled();
		}
	}

	/**
	 * Called when the dialog is created.
	 */
	@NonNull
	@Override
	public Dialog onCreateDialog(@Nullable Bundle savedInstanceState)
	{
		this.setupSharedPreferences();

		NacSharedConstants cons = this.getSharedConstants();

		return new AlertDialog.Builder(requireContext())
			.setTitle(cons.getTitleTextToSpeech())
			.setPositiveButton(cons.getActionOk(), (dialog, which) ->
				this.callOnTextToSpeechOptionsSelectedListener())
			.setNegativeButton(cons.getActionCancel(), (dialog, which) -> {})
			.setView(R.layout.dlg_alarm_text_to_speech)
			.create();
	}

	/**
	 */
	@Override
	public void onResume()
	{
		super.onResume();

		// Initialize the widgets
		AlertDialog dialog = (AlertDialog) getDialog();
		RelativeLayout useTtsContainer = dialog.findViewById(R.id.should_use_tts);

		// Set the member variable widgets
		this.mShouldUseTtsCheckBox = dialog.findViewById(R.id.should_use_tts_checkbox);
		this.mShouldUseTtsSummary = dialog.findViewById(R.id.should_use_tts_summary);
		this.mTtsFrequencyPicker = dialog.findViewById(R.id.tts_frequency_picker);

		// Setup the dialog and widgets
		useTtsContainer.setOnClickListener(this);
		this.setupShouldUseTts();
		this.setupTtsFrequencyPicker();
		this.setupTtsFrequencyEnabled();
		this.setupShouldUseTtsColor();
	}

	/**
	 * Set the default flag indicating whether text-to-speech should be used or
	 * not.
	 *
	 * @param  useTts  The default flag indicating whether text-to-speech should
	 *     be used or not
	 */
	public void setDefaultUseTts(boolean useTts)
	{
		this.mDefaultUseTts = useTts;
	}

	/**
	 * Set the default text-to-speech frequency.
	 *
	 * @param  freq  The default text-to-speech frequency.
	 */
	public void setDefaultTtsFrequency(int freq)
	{
		this.mDefaultTtsFrequency = freq;
	}

	/**
	 * Set the OnTextToSpeechOptionsSelectedListener object.
	 *
	 * @param  listener  The OnTextToSpeechOptionsSelectedListener object.
	 */
	public void setOnTextToSpeechOptionsSelectedListener(
		OnTextToSpeechOptionsSelectedListener listener)
	{
		this.mOnTextToSpeechOptionsSelectedListener = listener;
	}

	/**
	 * Setup the summary text for whether text-to-speech should be used or not.
	 */
	private void setupShouldUseTts()
	{
		boolean useTts = this.getDefaultUseTts();

		this.getShouldUseTtsCheckBox().setChecked(useTts);
		this.setupShouldUseTtsSummary();
	}

	/**
	 * Setup the summary text for whether text-to-speech should be used or not.
	 */
	private void setupShouldUseTtsColor()
	{
		NacSharedPreferences shared = this.getSharedPreferences();

		int[] colors = new int[] { shared.getThemeColor(), Color.GRAY };
		int[][] states = new int[][] {
			new int[] {  android.R.attr.state_checked },
			new int[] { -android.R.attr.state_checked } };
		ColorStateList colorStateList = new ColorStateList(states, colors);

		this.getShouldUseTtsCheckBox().setButtonTintList(colorStateList);
	}

	/**
	 * Setup the summary text for whether text-to-speech should be used or not.
	 */
	private void setupShouldUseTtsSummary()
	{
		boolean useTts = this.shouldUseTts();
		int textId = useTts ? R.string.speak_to_me_true : R.string.speak_to_me_false;

		this.getShouldUseTtsSummary().setText(textId);
	}

	/**
	 * Setup whether the text-to-speech frequency container can be used or not.
	 */
	private void setupTtsFrequencyEnabled()
	{
		NumberPicker picker = this.getTtsFrequencyPicker();
		boolean useTts = this.shouldUseTts();

		picker.setAlpha(useTts ? 1.0f : 0.25f);
		picker.setEnabled(useTts);
	}

	/**
	 * Setup scrollable picker for the text-to-speech frequency.
	 */
	private void setupTtsFrequencyPicker()
	{
		NacSharedConstants cons = this.getSharedPreferences().getConstants();
		NumberPicker picker = this.getTtsFrequencyPicker();
		int freq = this.getDefaultTtsFrequency();
		List<String> values = cons.getTextToSpeechFrequency();

		picker.setMinValue(0);
		picker.setMaxValue(values.size()-1);
		picker.setDisplayedValues((String[])values.toArray());
		picker.setValue(freq);
	}

	/**
	 * Get whether text-to-speech is being used or not.
	 *
	 * @return True if text-to-speech is being used, and False otherwise.
	 */
	private boolean shouldUseTts()
	{
		return this.getShouldUseTtsCheckBox().isChecked();
	}

	/**
	 * Toggle whether text-to-speech is being used or not.
	 */
	private void toggleShouldUseTts()
	{
		boolean useTts = this.shouldUseTts();

		this.getShouldUseTtsCheckBox().setChecked(!useTts);
	}

}

File: app/src/main/java/com/nfcalarmclock/tts/NacTextToSpeechDialog.kt
package com.nfcalarmclock.tts

import android.app.AlertDialog
import android.app.Dialog
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Bundle
import android.widget.NumberPicker
import android.widget.TextView
import com.google.android.material.checkbox.MaterialCheckBox
import com.nfcalarmclock.R
import com.nfcalarmclock.view.dialog.NacDialogFragment
import org.w3c.dom.Text

/**
 * Text to speech dialog asking the user what TTS settings they want.
 */
class NacTextToSpeechDialog
	: NacDialogFragment()
{

	companion object
	{

		/**
		 * Tag for the class.
		 */
		const val TAG = "NacAlarmTextToSpeechDialog"

	}

	/**
	 * Listener for when text-to-speech options are selected.
	 */
	fun interface OnTextToSpeechOptionsSelectedListener
	{
		fun onTextToSpeechOptionsSelected(
			shouldSayCurrentTime: Boolean,
			shouldSayAlarmName: Boolean,
			ttsFreq: Int)
	}

	/**
	 * Default flag indicating whether the current time should be said or not.
	 */
	var defaultSayCurrentTime = false

	/**
	 * Default flag indicating whether the alarm name should be said or not.
	 */
	var defaultSayAlarmName = false

	/**
	 * Default text-to-speech frequency.
	 */
	var defaultTtsFrequency = 0

	/**
	 * Textview for the section title for the frequency.
	 */
	private var howOftenToSayTextView: TextView? = null

	/**
	 * Checkbox for setting whether the time should be said or not.
	 */
	private var currentTimeCheckBox: MaterialCheckBox? = null

	/**
	 * Checkbox for setting whether the alarm name should be said or not.
	 */
	private var alarmNameCheckBox: MaterialCheckBox? = null

	/**
	 * Scrollable picker to choose the text-to-speech frequency.
	 */
	private var frequencyPicker: NumberPicker? = null

	/**
	 * Listener for when a text-to-speech option and frequency is selected.
	 */
	var onTextToSpeechOptionsSelectedListener: OnTextToSpeechOptionsSelectedListener? = null

	/**
	 * Whether text-to-speech is being used or not.
	 */
	private val shouldUseTts: Boolean
		get() = currentTimeCheckBox!!.isChecked || alarmNameCheckBox!!.isChecked

	/**
	 * The alpha that views should have based on the should use text-to-speech
	 * flag.
	 */
	private val alpha: Float
		get() = if (shouldUseTts) 1.0f else 0.25f

	/**
	 * Called when the dialog is created.
	 */
	override fun onCreateDialog(savedInstanceState: Bundle?): Dialog
	{
		// Setup the shared preferences
		setupSharedPreferences()

		// Create the dialog
		return AlertDialog.Builder(requireContext())
			.setTitle(R.string.title_tts)
			.setPositiveButton(R.string.action_ok) { _, _ ->

				// Call the listener
				onTextToSpeechOptionsSelectedListener?.onTextToSpeechOptionsSelected(
					currentTimeCheckBox!!.isChecked,
					alarmNameCheckBox!!.isChecked,
					frequencyPicker!!.value)

			}
			.setNegativeButton(R.string.action_cancel, null)
			.setView(R.layout.dlg_alarm_text_to_speech)
			.create()
	}

	/**
	 * Called when the dialog is resumed.
	 */
	override fun onResume()
	{
		// Super
		super.onResume()

		// Initialize the widgets
		//val whatToSay = dialog!!.findViewById<TextView>(R.id.title_tts_what_to_say)
		howOftenToSayTextView = dialog!!.findViewById(R.id.title_tts_how_often_to_say)

		// Set the member variables
		currentTimeCheckBox = dialog!!.findViewById(R.id.say_current_time)
		alarmNameCheckBox = dialog!!.findViewById(R.id.say_alarm_name)
		frequencyPicker = dialog!!.findViewById(R.id.tts_frequency_picker)

		// Set the default checked status
		currentTimeCheckBox!!.isChecked = defaultSayCurrentTime
		alarmNameCheckBox!!.isChecked = defaultSayAlarmName

		// Setup the views
		//whatToSay.setTextColor(sharedPreferences!!.nameColor)
		//howOftenToSay.setTextColor(sharedPreferences!!.nameColor)
		setupCheckBoxColor()
		setupOnClickListener()
		setupFrequencyPickerValues()
		setupFrequencyPickerUsable()
	}

	/**
	 * Setup the color of the use text-to-speech checkbox.
	 */
	private fun setupCheckBoxColor()
	{
		// Get the colors for the boolean states
		val colors = intArrayOf(sharedPreferences!!.themeColor, Color.GRAY)

		// Get the IDs of the two states
		val states = arrayOf(intArrayOf(android.R.attr.state_checked),
			intArrayOf(-android.R.attr.state_checked))

		// Set the state list of the checkboxes
		currentTimeCheckBox!!.buttonTintList = ColorStateList(states, colors)
		alarmNameCheckBox!!.buttonTintList = ColorStateList(states, colors)
	}

	/**
	 * Setup whether the text-to-speech frequency picker can be used or not.
	 */
	private fun setupFrequencyPickerUsable()
	{
		// Set the alpha of the frequency title
		howOftenToSayTextView!!.alpha = alpha

		// Set the alpha of the frequency picker
		frequencyPicker!!.alpha = alpha

		// Set whether the frequency picker should be accessible or not
		frequencyPicker!!.isEnabled = shouldUseTts
	}

	/**
	 * Setup the values of the text-to-speech frequency picker.
	 */
	private fun setupFrequencyPickerValues()
	{
		// Get the max frequency size
		val values = requireContext().resources.getStringArray(R.array.tts_frequency).toList()

		// Setup the frequency picker
		frequencyPicker!!.minValue = 0
		frequencyPicker!!.maxValue = values.size - 1
		frequencyPicker!!.displayedValues = values.toTypedArray()
		frequencyPicker!!.value = defaultTtsFrequency
	}

	/**
	 * Setup the on click listeners.
	 */
	private fun setupOnClickListener()
	{
		// Set the on click listener for the current time checkbox container
		currentTimeCheckBox!!.setOnClickListener {

			// Enable/disable the frequency picker
			setupFrequencyPickerUsable()

		}

		// Set the on click listener for the current time checkbox container
		alarmNameCheckBox!!.setOnClickListener {

			// Enable/disable the frequency picker
			setupFrequencyPickerUsable()

		}
	}

}